code/controllers/subsystem/achievements.dm

```diff
@@ -0,0 +1,84 @@
+SUBSYSTEM_DEF(achievements)
+	name = "Achievements"
+	flags = SS_NO_FIRE
+	var/list/achievements = list()
+	var/list/cached_achievements = list()
+	var/list/browsers = list()
+	var/list/achievementsEarned = list()
+
+/datum/controller/subsystem/achievements/Initialize(timeofday)
+	for(var/i in subtypesof(/datum/achievement))
+		var/datum/achievement/A = i
+		achievements[A] = initial(A.id)
+
+		var/datum/DBQuery/medalQuery = SSdbcore.NewQuery("SELECT name, desc FROM [format_table_name("achievements")] WHERE id = '[initial(A.id)]'")
+		medalQuery.Execute()
+		if(!medalQuery.NextRow())
+			var/datum/DBQuery/medalQuery2 = SSdbcore.NewQuery("INSERT INTO [format_table_name("achievements")] (name, id, desc) VALUES ('[initial(A.name)]', '[initial(A.id)]', '[initial(A.desc)]')")
+			medalQuery2.Execute()
+			qdel(medalQuery2)
+		else if(medalQuery.item[1] != initial(A.name) || medalQuery.item[2] != initial(A.desc))
+			var/datum/DBQuery/medalQuery2 = SSdbcore.NewQuery("UPDATE [format_table_name("achievements")] SET name = '[initial(A.name)]', desc = '[initial(A.desc)]' WHERE id = '[initial(A.id)]'")
+			medalQuery2.Execute()
+			qdel(medalQuery2)
+		
+		qdel(medalQuery)
+		
+	
+	var/datum/DBQuery/ridOldChieves = SSdbcore.NewQuery("SELECT id FROM [format_table_name("achievements")]")
+	ridOldChieves.Execute()
+	while(ridOldChieves.NextRow())
+		var/id = ridOldChieves.item[1]
+		var/found_achievement = FALSE
+		for(var/A in achievements)
+			if(achievements[A] == id)
+				found_achievement = TRUE
+				break
+		if(!found_achievement)
+			stack_trace("Old achievement [id] found in database, removing")
+			var/datum/DBQuery/getRidOfOldStuff = SSdbcore.NewQuery("DELETE FROM [format_table_name("achievements")] WHERE id = '[id]'")
+			getRidOfOldStuff.Execute()
+			var/datum/DBQuery/ridTheOtherTableAsWell = SSdbcore.NewQuery("DELETE FROM [format_table_name("earned_achievements")] WHERE id = '[id]'")
+			ridTheOtherTableAsWell.Execute()
+			qdel(ridTheOtherTableAsWell)
+			qdel(getRidOfOldStuff)
+
+	qdel(ridOldChieves)
+	return ..()
+
+/datum/controller/subsystem/achievements/proc/unlock_achievement(datum/achievement/achievement, client/C)
+	if(!achievements[achievement])
+		stack_trace("Achievement [initial(achievement.name)] not found in list of achievements when trying to unlock for [C.ckey]")
+		return FALSE
+	if(!has_achievement(achievement, C))
+		var/datum/DBQuery/medalQuery = SSdbcore.NewQuery("INSERT INTO [format_table_name("earned_achievements")] (ckey, id) VALUES ('[C.ckey]', '[achievements[achievement]]')")
+		medalQuery.Execute()
```

Comment:
[13:31:13] Runtime in achievements.dm, line 55: Cannot execute 0.Execute().
proc name: unlock achievement (/datum/controller/subsystem/achievements/proc/unlock_achievement)
usr: Nichlas0010/(Maggie May)
usr.loc: (Lavaland Wastes (102, 10, 5))
src: Achievements (/datum/controller/subsystem/achievements)
call stack:
Achievements (/datum/controller/subsystem/achievements): unlock achievement(/datum/achievement/bubblegum (/datum/achievement/bubblegum), Nichlas0010 (/client))
world: WrapAdminProcCall(Achievements (/datum/controller/subsystem/achievements), "unlock_achievement", /list (/list))
WrapAdminProcCall(Achievements (/datum/controller/subsystem/achievements), "unlock_achievement", /list (/list))
Nichlas0010 (/client): Atom ProcCall(Achievements (/datum/controller/subsystem/achievements))